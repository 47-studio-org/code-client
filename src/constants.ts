import { CustomDCIgnore, DefaultDCIgnore } from '@deepcode/dcignore';

export const MAX_PAYLOAD = 4 * 1024 * 1024;
export const defaultBaseURL = 'https://www.deepcode.ai';
export const apiPath = '/publicapi';
export const HASH_ALGORITHM = 'sha256';
export const ENCODE_TYPE = 'hex';
export const GIT_FILENAME = '.git';
export const GITIGNORE_FILENAME = '.gitignore';
export const DCIGNORE_FILENAME = '.dcignore';
export const EXCLUDED_NAMES = [GIT_FILENAME, GITIGNORE_FILENAME, DCIGNORE_FILENAME];

export const IGNORES_DEFAULT = [`**/${GIT_FILENAME}`];

export const IGNORE_FILES_NAMES = [GITIGNORE_FILENAME, DCIGNORE_FILENAME];

export const DCIGNORE_DRAFTS = {
  custom: CustomDCIgnore,
  default: DefaultDCIgnore,
};

// // eslint-disable-next-line no-shadow
// export enum FILE_CURRENT_STATUS {
//   modified = 'modified',
//   deleted = 'deleted',
//   same = 'same',
//   created = 'created',
// }

// eslint-disable-next-line no-shadow
export enum RequestTypes {
  startSession = 'startSession',
  checkSession = 'checkSession',
  getFilters = 'getFilters',
  createBundle = 'createBundle',
  checkBundle = 'checkBundle',
  extendBundle = 'extendBundle',
  uploadFiles = 'uploadFiles',
  getAnalysis = 'getAnalysis',
  reportError = 'reportError',
  reportEvent = 'reportEvent',
}

export const ERROR_CODES = new Set([
  304, // loginInProgress,
  400, // unauthorizedContent,
  401, // unauthorizedUser,
  403, // unauthorizedBundleAccess,
  404, // notFound,
  413, // bigPayload,
  500, // serverError,
  502, // badGateway,
  503, // serviceUnavailable,
  504, // timeout,
]);

export const ERRORS = {
  [RequestTypes.startSession]: {
    other: 'Login failed',
  },
  [RequestTypes.checkSession]: {
    401: 'Missing or invalid sessionToken',
    other: 'Checking session failed',
  },
  [RequestTypes.getFilters]: {
    401: 'Missing sessionToken or incomplete login process',
    other: 'Getting filters failed',
  },
  [RequestTypes.createBundle]: {
    400: "Request content doesn't match the specifications",
    401: 'Missing sessionToken or incomplete login process',
    403: 'Unauthorized access to requested repository',
    404: 'Unable to resolve requested oid',
    413: 'Payload too large',
    other: 'Creating bundle failed',
  },
  [RequestTypes.checkBundle]: {
    401: 'Missing sessionToken or incomplete login process',
    403: 'Unauthorized access to requested bundle',
    404: 'Uploaded bundle has expired',
    other: 'Checking bundle failed',
  },
  [RequestTypes.extendBundle]: {
    400: 'Invalid request, attempted to extend a git bundle, or ended up with an empty bundle after the extension',
    401: 'Missing sessionToken or incomplete login process',
    403: 'Unauthorized access to parent bundle',
    404: 'Parent bundle has expired',
    413: 'Payload too large',
    other: 'Extending bundle failed',
  },
  [RequestTypes.uploadFiles]: {
    400: 'Invalid request, attempted to extend a git bundle, or ended up with an empty bundle after the extension',
    401: 'Missing sessionToken or incomplete login process',
    403: 'Unauthorized access to parent bundle',
    404: 'Parent bundle has expired',
    413: 'Payload too large',
    other: 'Uploading files failed',
  },
  [RequestTypes.getAnalysis]: {
    401: 'Missing sessionToken or incomplete login process',
    403: 'Unauthorized access to requested bundle',
    other: 'Getting analysis failed',
  },
  [RequestTypes.reportError]: {
    other: 'Reporting error failed',
  },
  [RequestTypes.reportEvent]: {
    other: 'Reporting event failed',
  },
};

export const BUNDLE_ERRORS = {
  create: {
    400: 'Creating bundle: Request content does not match the specifications',
    401: 'Creating bundle: Missing sessionToken or incomplete login process',
    403: 'Creating bundle: Unauthorized access to requested repository',
    413: 'Creating bundle: Payload too large',
  },
  upload: {
    400: 'Uploading files: Content and hash mismatch or attempted to upload files to a git bundle',
    401: 'Uploading files: Missing sessionToken or incomplete login process',
    403: 'Uploading files: Unauthorized access to requested bundle',
    413: 'Uploading files: Payload too large',
  },
  check: {
    401: 'Check bundle: Missing sessionToken or incomplete login process',
    403: 'Check bundle: Unauthorized access to requested bundle',
    404: 'Check bundle: Uploaded bundle has expired',
  },
  extend: {
    400: 'Extending bundle: Attempted to extend a git bundle, or ended up with an empty bundle after the extension',
    401: 'Extending bundle: Missing sessionToken or incomplete login process',
    403: 'Extending bundle: Unauthorized access to parent bundle',
    404: 'Extending bundle: Parent bundle has expired',
    413: 'Extending bundle: Payload too large',
  },
};
