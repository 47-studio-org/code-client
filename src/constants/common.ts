import path from 'path';

export const isWindows = path.sep === '\\';

// Max payload is 5 MB, but we will use 4 MB for files due to we need also to send
// another info about files with their contents
export const maxPayload = 4 * 1024 * 1024;

export const defaultBaseURL = 'https://www.deepcode.ai';
export const apiPath = '/publicapi';
