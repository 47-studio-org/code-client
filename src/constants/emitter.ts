export const CUSTOM_EVENTS = {
  buildBundleProgress: 'buildBundleProgress',
  buildBundleFinish: 'buildBundleFinish',
  createBundleProgress: 'createBundleProgress',
  createBundleFinish: 'createBundleFinish',
  uploadBundleProgress: 'uploadBundleProgress',
  uploadFilesFinish: 'uploadFilesFinish',
  analyseProgress: 'analyseProgress',
  analyseFinish: 'analyseFinish',
};
