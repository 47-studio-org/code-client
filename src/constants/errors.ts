import { RequestTypes } from '../enums/request-types.enum';

export const ERROR_CODES = new Set([
  304, // loginInProgress,
  400, // unauthorizedContent,
  401, // unauthorizedUser,
  403, // unauthorizedBundleAccess,
  404, // notFound,
  413, // bigPayload,
  500, // serverError,
  502, // badGateway,
  503, // serviceUnavailable,
  504, // timeout,
]);

export const ERRORS = {
  [RequestTypes.startSession]: {
    other: 'Login failed',
  },
  [RequestTypes.checkSession]: {
    401: 'Missing or invalid sessionToken',
    other: 'Checking session failed',
  },
  [RequestTypes.getFilters]: {
    401: 'Missing sessionToken or incomplete login process',
    other: 'Getting filters failed',
  },
  [RequestTypes.createBundle]: {
    400: "Request content doesn't match the specifications",
    401: 'Missing sessionToken or incomplete login process',
    403: 'Unauthorized access to requested repository',
    404: 'Unable to resolve requested oid',
    413: 'Payload too large',
    other: 'Creating bundle failed',
  },
  [RequestTypes.checkBundle]: {
    401: 'Missing sessionToken or incomplete login process',
    403: 'Unauthorized access to requested bundle',
    404: 'Uploaded bundle has expired',
    other: 'Checking bundle failed',
  },
  [RequestTypes.extendBundle]: {
    400: 'Invalid request, attempted to extend a git bundle, or ended up with an empty bundle after the extension',
    401: 'Missing sessionToken or incomplete login process',
    403: 'Unauthorized access to parent bundle',
    404: 'Parent bundle has expired',
    413: 'Payload too large',
    other: 'Extending bundle failed',
  },
  [RequestTypes.uploadFiles]: {
    400: 'Invalid request, attempted to extend a git bundle, or ended up with an empty bundle after the extension',
    401: 'Missing sessionToken or incomplete login process',
    403: 'Unauthorized access to parent bundle',
    404: 'Parent bundle has expired',
    413: 'Payload too large',
    other: 'Uploading files failed',
  },
  [RequestTypes.getAnalysis]: {
    401: 'Missing sessionToken or incomplete login process',
    403: 'Unauthorized access to requested bundle',
    other: 'Getting analysis failed',
  },
  [RequestTypes.reportError]: {
    other: 'Reporting error failed',
  },
  [RequestTypes.reportEvent]: {
    other: 'Reporting event failed',
  },
};
