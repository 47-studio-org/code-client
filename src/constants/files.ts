export const CRYPTO = {
  algorithm: 'sha256',
  hashEncode: 'hex',
};

export const HASH_ALGORITHM = 'sha256';
export const ENCODE_TYPE = 'hex';
export const GIT_FILENAME = '.git';
export const GITIGNORE_FILENAME = '.gitignore';
export const DCIGNORE_FILENAME = '.dcignore';
export const EXCLUDED_NAMES = [GIT_FILENAME, GITIGNORE_FILENAME, DCIGNORE_FILENAME];
export const ALLOWED_PAYLOAD_SIZE = 1024 * 1024 * 4; // max payload size of 4MB in bytes

export const FILE_CURRENT_STATUS = {
  modified: 'modified',
  deleted: 'deleted',
  same: 'same',
  created: 'created',
};
