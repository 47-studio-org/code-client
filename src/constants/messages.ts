export const BUNDLE_ERRORS = {
  create: {
    400: 'Creating bundle: Request content does not match the specifications',
    401: 'Creating bundle: Missing sessionToken or incomplete login process',
    403: 'Creating bundle: Unauthorized access to requested repository',
    413: 'Creating bundle: Payload too large',
  },
  upload: {
    400: 'Uploading files: Content and hash mismatch or attempted to upload files to a git bundle',
    401: 'Uploading files: Missing sessionToken or incomplete login process',
    403: 'Uploading files: Unauthorized access to requested bundle',
    413: 'Uploading files: Payload too large',
  },
  check: {
    401: 'Check bundle: Missing sessionToken or incomplete login process',
    403: 'Check bundle: Unauthorized access to requested bundle',
    404: 'Check bundle: Uploaded bundle has expired',
  },
  extend: {
    400: 'Extending bundle: Attempted to extend a git bundle, or ended up with an empty bundle after the extension',
    401: 'Extending bundle: Missing sessionToken or incomplete login process',
    403: 'Extending bundle: Unauthorized access to parent bundle',
    404: 'Extending bundle: Parent bundle has expired',
    413: 'Extending bundle: Payload too large',
  },
};