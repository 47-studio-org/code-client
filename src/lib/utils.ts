// We are using the same types that Object.fromEntries has
// eslint-disable-next-line import/prefer-default-export, @typescript-eslint/no-explicit-any
export function fromEntries(entries: Iterable<readonly any[]>): any;
export function fromEntries<T = any>(
  entries: Iterable<readonly [PropertyKey, T]>,
): {
  [k in PropertyKey]: T;
} {
  return [...entries].reduce((obj, [key, val]) => {
    obj[key] = val; // eslint-disable-line no-param-reassign
    return obj;
  }, {});
}
