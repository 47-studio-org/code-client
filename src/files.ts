import * as nodePath from 'path';
import * as fs from 'fs';
import fg from 'fast-glob';
import crypto, { HexBase64Latin1Encoding } from 'crypto';
import { union } from 'lodash';
import util from 'util';
import flatCache from 'flat-cache';

import { HASH_ALGORITHM, ENCODE_TYPE, MAX_PAYLOAD, IGNORES_DEFAULT, IGNORE_FILES_NAMES } from './constants';

import { ISupportedFiles, IFileInfo } from './interfaces/files.interface';

const isWindows = nodePath.sep === '\\';

const lStat = util.promisify(fs.lstat);
const readFile = util.promisify(fs.readFile);

type CachedData = [number, number, string];

export function isFileSupported(path: string, supportedFiles: ISupportedFiles): boolean {
  return supportedFiles.configFiles.includes(path) || supportedFiles.extensions.includes(nodePath.extname(path));
}

export function parseFileIgnores(path: string): string[] {
  const dirname = nodePath.dirname(path);
  const f = fs.readFileSync(path, { encoding: 'utf8' });

  const rules = f
    .split('\n')
    .map(l => l.trim().replace(/\/$/, '')) // Remove white spaces and trim slashes
    .filter(l => !!l && !l.startsWith('#'));

  const results: string[] = [];
  for (const rule of rules) {
    results.push(nodePath.posix.join(dirname, rule));
    if (!rule.startsWith('/')) {
      results.push(nodePath.posix.join(dirname, '**', rule));
    }
  }
  return results;
}

export function getGlobPatterns(supportedFiles: ISupportedFiles): string[] {
  return [...supportedFiles.extensions.map(e => `*${e}`), ...supportedFiles.configFiles];

  // return `**/\{${patterns.join(',')}\}`;
}

function* scanDir(
  path: string,
  globPatterns: string[],
  symlinksEnabled = false,
  fileIgnores: string[] = IGNORES_DEFAULT,
): Generator<fg.Entry> {
  let localFileIgnores = [...fileIgnores];

  // Check ignore files inside this directory
  const localIgnoreFiles = fg.sync(IGNORE_FILES_NAMES, {
    dot: true,
    absolute: true,
    cwd: path,
    deep: 1,
    ignore: localFileIgnores,
    caseSensitiveMatch: false,
    followSymbolicLinks: symlinksEnabled,
    onlyFiles: true,
    objectMode: false,
    stats: false,
  });

  // Read ignore files and merge new patterns
  const newIgnorePatterns = localIgnoreFiles.map(p => {
    console.debug('recognized ignore rules in file --> ', p);
    return parseFileIgnores(p);
  });
  if (newIgnorePatterns.length) {
    localFileIgnores = union(localFileIgnores, ...newIgnorePatterns);
  }

  // Scan files
  const localFiles = fg.sync(globPatterns, {
    dot: true,
    absolute: true,
    cwd: path,
    deep: 1,
    ignore: localFileIgnores,
    caseSensitiveMatch: false,
    followSymbolicLinks: symlinksEnabled,
    onlyFiles: true,
    objectMode: true,
    stats: true,
  });

  for (const f of localFiles) {
    yield f;
  }

  // Scan sub-directories
  const subDirs = fg.sync('**', {
    onlyDirectories: true,
    dot: true,
    absolute: true,
    cwd: path,
    deep: 1,
    ignore: localFileIgnores,
    caseSensitiveMatch: false,
    followSymbolicLinks: symlinksEnabled,
    objectMode: false,
    stats: false,
  });

  for (const d of subDirs) {
    for (const f of scanDir(d, globPatterns, symlinksEnabled, localFileIgnores)) {
      yield f;
    }
  }
}

export function determineBaseDir(paths: string[]): string {
  if (paths.length) {
    const path = paths[0];
    const stats = fs.lstatSync(path);
    if (stats.isFile()) {
      return nodePath.dirname(path);
    }

    return path;
  }
  return '';
}

/**
 * Returns bundle files from requested paths
 * */
export async function* collectBundleFiles(
  baseDir: string,
  paths: string[],
  supportedFiles: ISupportedFiles,
  maxFileSize = MAX_PAYLOAD,
  symlinksEnabled = false,
  fileIgnores: string[] = IGNORES_DEFAULT,
): AsyncGenerator<IFileInfo> {
  const globPatterns = getGlobPatterns(supportedFiles);
  // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-member-access
  const cache = flatCache.load('.dccache', baseDir);

  for (const path of paths) {
    // eslint-disable-next-line no-await-in-loop
    const fileStats = await lStat(path);
    // Check if symlink and exclude if requested
    if (fileStats.isSymbolicLink() && !symlinksEnabled) continue;

    if (fileStats.isFile() && isFileSupported(path, supportedFiles)) {
      if (fileStats.size > maxFileSize) continue;
      yield getFileInfo(path, baseDir, false, cache);
    } else if (fileStats.isDirectory()) {
      for (const entry of scanDir(path, globPatterns, symlinksEnabled, fileIgnores)) {
        if (entry.stats && entry.stats.size > maxFileSize) continue;

        yield getFileInfo(entry.path, baseDir, false, cache);
      }
    }
  }

  // eslint-disable-next-line @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-member-access
  cache.save();
}

export async function getFileInfo(
  filePath: string,
  baseDir: string,
  withContent = false,
  cache: flatCache.Cache | null = null,
): Promise<IFileInfo> {
  const fileStats = await lStat(filePath);

  const relPath = nodePath.relative(baseDir, filePath);

  const calcHash = (content: string) => {
    return crypto
      .createHash(HASH_ALGORITHM)
      .update(content)
      .digest(ENCODE_TYPE as HexBase64Latin1Encoding);
  };

  let fileContent = '';
  let fileHash = '';
  if (!withContent && !!cache) {
    // Try to get hash from cache
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-member-access
    const cachedData: CachedData | null = cache.getKey(relPath);
    if (cachedData) {
      if (cachedData[0] === fileStats.size && cachedData[1] === fileStats.mtimeMs) {
        // eslint-disable-next-line prefer-destructuring
        fileHash = cachedData[2];
      } else {
        // console.log(`did not match cache for: ${relPath} | ${cachedData} !== ${[fileStats.size, fileStats.mtime]}`);
      }
    }
  }

  if (!fileHash) {
    fileContent = await readFile(filePath, 'utf8');
    fileHash = calcHash(fileContent);
    // eslint-disable-next-line @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-member-access
    cache?.setKey(relPath, [fileStats.size, fileStats.mtimeMs, fileHash]);
  }

  const posixPath = !isWindows ? relPath : relPath.replace('\\', '/');

  return {
    filePath,
    bundlePath: `/${posixPath}`,
    size: fileStats.size,
    hash: fileHash,
    content: withContent ? fileContent : undefined,
  };
}

export async function resolveBundleFiles(baseDir: string, bundleMissingFiles: string[]): Promise<IFileInfo[]> {
  // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-member-access
  const cache = flatCache.load('.dccache', baseDir);
  const tasks = bundleMissingFiles.map(mf => {
    const filePath = resolveBundleFilePath(baseDir, mf);
    return getFileInfo(filePath, baseDir, true, cache);
  });

  const res = await Promise.all(tasks);
  // eslint-disable-next-line @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-member-access
  cache.save(true);
  return res;
}

export function resolveBundleFilePath(baseDir: string, bundleFilePath: string): string {
  let relPath = bundleFilePath.slice(1);

  if (isWindows) {
    relPath = relPath.replace('/', '\\');
  }

  return nodePath.resolve(baseDir, relPath);
}

export function* composeFilePayloads(files: IFileInfo[], bucketSize = MAX_PAYLOAD): Generator<IFileInfo[]> {
  type Bucket = {
    size: number;
    files: IFileInfo[];
  };
  const buckets: Bucket[] = [{ size: bucketSize, files: [] }];

  let bucketIndex = -1;
  const isLowerSize = (bucket: Bucket, fileData: IFileInfo) => bucket.size >= fileData.size;
  for (let fileData of files) {
    if (fileData.size > bucketSize) {
      // This file is too large. but it should not be here as previosly checked
      fileData = { ...fileData, size: 1, content: '' };
    }

    // Find suitable bucket
    bucketIndex = buckets.findIndex(b => isLowerSize(b, fileData));

    if (bucketIndex === -1) {
      // Create a new bucket
      buckets.push({ size: bucketSize, files: [] });
      bucketIndex = buckets.length - 1;
    }

    buckets[bucketIndex].files.push(fileData);
    buckets[bucketIndex].size -= fileData.size;

    if (buckets[bucketIndex].size < bucketSize * 0.01) {
      yield buckets[bucketIndex].files; // Give bucket to requester
      buckets.splice(bucketIndex); // Remove it as fullfilled
    }
  }

  // Send all left-over buckets
  for (const bucket of buckets.filter(b => b.files.length)) {
    yield bucket.files;
  }
}
