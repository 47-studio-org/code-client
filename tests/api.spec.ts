import { baseURL, authHost, sessionToken, TEST_TIMEOUT } from './constants/base';
import { bundleFiles, bundleFilesFull, bundleFilePaths } from './constants/sample';
import { fromEntries } from '../src/lib/utils';
import pick from 'lodash.pick';
import {
  getFilters,
  startSession,
  checkSession,
  createBundle,
  checkBundle,
  extendBundle,
  getAnalysis,
  AnalysisStatus,
} from '../src/http';
import { BundleFiles } from '../src/interfaces/files.interface';

const fakeBundleHash = 'fe72b6d08bbbc36214992aa769a9e3ee2a52579f88e1212cbb0719d63da6e14c';
let fakeBundleHashFull = '';
const realBundleHash = '26635e09233a4c221b2d0f3f118f92e0b48bbdcbbbcde8e3d0d34d1eacb44b37';
let realBundleHashFull = '';

const fakeMissingFiles = [
  '/AnnotatorTest.cpp',
  '/GitHubAccessTokenScrambler12.java',
  '/app.js',
  '/db.js',
  '/main.js',
  // TODO: This should be ignored
  '/not/ignored/this_should_be_ignored.jsx',
  '/not/ignored/this_should_not_be_ignored.java',
  '/routes/index.js',
  '/routes/sharks.js',
];

describe('Requests to public API', () => {
  it('gets filters successfully', async () => {
    const response = await getFilters(baseURL, '');
    expect(response.type).toEqual('success');
    if (response.type === 'error') return;
    expect(new Set(response.value.configFiles)).toEqual(new Set(['.dcignore', '.gitignore']));
    expect(new Set(response.value.extensions)).toEqual(
      new Set([
        '.CS',
        '.Cs',
        '.c',
        '.cc',
        '.cpp',
        '.cs',
        '.cxx',
        '.es',
        '.es6',
        '.h',
        '.hpp',
        '.htm',
        '.html',
        '.hxx',
        '.java',
        '.js',
        '.jsx',
        '.php',
        '.py',
        '.ts',
        '.tsx',
        '.vue',
      ]),
    );

    expect(response.value.configFiles.length).toBeGreaterThan(0);
    expect(response.value.extensions.length).toBeGreaterThan(0);
  });

  it('starts session successfully', async () => {
    const startSessionResponse = startSession({
      source: 'atom',
      authHost,
    });
    expect(startSessionResponse.loginURL).toMatch(/.*\/login\?token=.*&utm_source=atom/);
    const draftToken = startSessionResponse.draftToken;

    // This token is just a draft and not ready to be used permanently
    const checkSessionResponse = await checkSession({ authHost, draftToken });
    expect(checkSessionResponse.type).toEqual('success');
    if (checkSessionResponse.type == 'error') return;
    expect(checkSessionResponse.value).toEqual('');
  });

  it('checks session unsuccessfully', async () => {
    const response = await checkSession({
      authHost,
      draftToken: 'dummy-token',
    });
    expect(response.type).toEqual('success');
    if (response.type === 'error') return;
    expect(response.value).toEqual('');
  });

  // TODO: find a way to test successfull workflow automatically
  // it('checks session successfully', async () => {
  //   const response = await checkSession({
  //     authHost,
  //     sessionToken,
  //   });
  //   expect(response.type).toEqual('success');
  //   if (response.type === 'error') return;
  //   expect(response.value).toEqual(true);
  // });

  it(
    'creates bundle successfully',
    async () => {

      const files: BundleFiles = fromEntries([...(await bundleFiles).entries()].map(([i, d]) => [d.bundlePath, `${i}`]));

      const response = await createBundle({
        baseURL,
        sessionToken,
        files,
        source: 'atom',
      });
      expect(response.type).toEqual('success');
      if (response.type === 'error') {
        console.error(response);
        return;
      }
      expect(response.value.bundleHash).toContain(fakeBundleHash);
      fakeBundleHashFull = response.value.bundleHash;
      expect(response.value.missingFiles).toEqual(fakeMissingFiles);
    },
    TEST_TIMEOUT,
  );

  it(
    'checks bundle successfully',
    async () => {
      const response = await checkBundle({
        baseURL,
        sessionToken,
        bundleHash: fakeBundleHashFull,
      });
      expect(response.type).toEqual('success');
      if (response.type === 'error') return;
      expect(response.value.bundleHash).toEqual(fakeBundleHashFull);
      expect(response.value.missingFiles).toEqual(fakeMissingFiles);
    },
    TEST_TIMEOUT,
  );

  it(
    'checks expired bundle successfully',
    async () => {
      const response = await checkBundle({
        baseURL,
        sessionToken,
        bundleHash: 'mock-expired-bundle-id',
      });
      expect(response.type).toEqual('error');
      // dummy to cheat typescript compiler
      if (response.type == 'success') return;
      expect(response.error.statusCode).toEqual(404);
      expect(response.error.statusText).toEqual('Uploaded bundle has expired');
    },
    TEST_TIMEOUT,
  );

  it(
    'request analysis with missing files',
    async () => {
      let response;
      do {
        response = await getAnalysis({
          baseURL,
          sessionToken,
          bundleHash: fakeBundleHashFull,
          severity: 1,
          source: 'atom',
        });
      } while (response.type === 'success');

      expect(response.type).toEqual('error');
      expect(response.error).toEqual({
        apiName: 'getAnalysis',
        statusCode: 404,
        statusText: 'Not found',
      });
    },
    TEST_TIMEOUT,
  );

  it(
    'extends bundle successfully',
    async () => {
      const response = await extendBundle({
        baseURL,
        sessionToken,
        bundleHash: fakeBundleHashFull,
        files: {
          '/new.js': 'new123',
        },
        removedFiles: [
          `/AnnotatorTest.cpp`,
          `/app.js`,
          `/GitHubAccessTokenScrambler12.java`,
          `/db.js`,
          `/main.js`,
          `/not/ignored/this_should_be_ignored.jsx`,
          `/not/ignored/this_should_not_be_ignored.java`,
          `/routes/index.js`,
          `/routes/sharks.js`,
        ],
      });
      expect(response.type).toEqual('success');
      if (response.type === 'error') return;
      expect(response.value.bundleHash).toContain('f497974a35b2b146bf35c259186027d71fdecdeb319567e46255090edca06675');
      expect(response.value.missingFiles).toEqual([`/new.js`]);
    },
    TEST_TIMEOUT,
  );

  it(
    'extends expired bundle successfully',
    async () => {
      const response = await extendBundle({
        baseURL,
        sessionToken,
        bundleHash: 'wrong-bundle-id-2',
        files: {
          '/new2.js': 'new1234',
        },
      });

      expect(response.type).toEqual('success');
      if (response.type === 'error') return;
      expect(response.value.bundleHash).toContain('5979992527b34d10f9d772abcc0d6a07c70db08b4850b1c99ee6879f14d13f22');
      expect(response.value.missingFiles).toEqual([`/new2.js`]);
    },
    TEST_TIMEOUT,
  );

  it(
    'uploads fake files to fake bundle',
    async () => {
      const response = await extendBundle({
        baseURL,
        sessionToken,
        bundleHash: fakeBundleHashFull,
        files: {
          'df.js': { hash: 'df', content: 'const module = new Module();'},
          'sdfs.js': { hash: 'sdfs', content: 'const App = new App();'},
        },
      });
      expect(response.type).toEqual('success');
      if (response.type !== 'success') return; // TS trick
      expect(response.value.bundleHash).toContain('9ef5c6239fc95b9dc4e9defb9550dee534580cc990cee773f0c504688e86cac8');
      expect(response.value.missingFiles).toHaveLength(11);
    },
    TEST_TIMEOUT,
  );

  it('test successful workflow', async () => {
      // Create a bundle first
      const files: BundleFiles = (await bundleFilesFull).reduce((r, d) => {
        r[d.bundlePath] = pick(d, ['hash', 'content']);
        return r;
      }, {});

      const bundleResponse = await createBundle({
        baseURL,
        sessionToken,
        files,
        source: 'atom',
      });
      expect(bundleResponse.type).toEqual('success');
      if (bundleResponse.type === 'error') return;
      expect(bundleResponse.value.bundleHash).toContain(realBundleHash);
      realBundleHashFull = bundleResponse.value.bundleHash;

      // Check missing files
      const checkResponse = await checkBundle({
        baseURL,
        sessionToken,
        bundleHash: realBundleHashFull,
      });
      expect(checkResponse.type).toEqual('success');
      if (checkResponse.type === 'error') return;
      expect(checkResponse.value.bundleHash).toEqual(realBundleHashFull);
      expect(checkResponse.value.missingFiles).toEqual([]);

      // Get analysis results
      let response = await getAnalysis({
        baseURL,
        sessionToken,
        bundleHash: realBundleHashFull,
        severity: 1,
        source: 'atom',
      });
      expect(response.type).toEqual('success');
      if (response.type === 'error') return;
      expect(response.value.status !== AnalysisStatus.failed).toBeTruthy();

      if (response.value.status === AnalysisStatus.complete) {

        const suggestions = response.value.sarif.runs[0].results!;
        expect(suggestions).toHaveLength(12);
        if (suggestions.length < 1) return; // TS trick

        const suggestion = suggestions[0];
        expect(suggestion).toMatchObject({
          ruleId: 'cpp/MissingOpenCheckOnFile/test',
          level: 'note',
          fingerprints: {
            '0': '3e40a81739245db8fff4903a7e28e08bffa03486a677e7c91594cfdf15fb5a1d',
            '1': '57664a44.2c254dac.98501263.9e345555.da547a36.9509b717.a713c1c8.45d76bdf.4a7ae834.2c254dac.98501263.9e345555.da547a36.9509b717.a713c1c8.45d76bdf'
          },
          message: {
            "text": "Missing check is_open on {0} before {1}.",
            "markdown": "Missing check is_open on std::fstream before writing to it.",
            "arguments": [
              "[std::fstream](0)",
              "[writing to it](1)",
            ]
          }
        });

        expect(suggestion.codeFlows![0].threadFlows[0].locations).toHaveLength(2);

        expect(new Set(response.value.coverage)).toEqual(
          new Set([
            {
              files: 1,
              isSupported: true,
              lang: 'C++ (beta)',
            },
            {
              files: 2,
              isSupported: true,
              lang: 'Java',
            },
            {
              files: 1,
              isSupported: true,
              lang: 'C++ (beta)',
            },
            {
              files: 5,
              isSupported: true,
              lang: 'JavaScript',
            },
            {
              files: 1,
              isSupported: true,
              lang: 'JSX',
            },
          ]),
        );
      }

      // Get analysis results limited to 1 file
      do {
        response = await getAnalysis({
          baseURL,
          sessionToken,
          bundleHash: realBundleHashFull,
          severity: 1,
          limitToFiles: [`/GitHubAccessTokenScrambler12.java`],
          source: 'atom',
        });

        expect(response.type).toEqual('success');
        if (response.type === 'error') return;
        expect(response.value.status !== AnalysisStatus.failed).toBeTruthy();
      } while (response.value.status !== AnalysisStatus.complete);

      expect(response.value.sarif.runs[0].results).toHaveLength(8);

      // Get analysis results with severity 3
      do {
        response = await getAnalysis({
          baseURL,
          sessionToken,
          bundleHash: realBundleHashFull,
          severity: 3,
          source: 'atom',
        });
        expect(response.type).toEqual('success');
        if (response.type === 'error') return;
        expect(response.value.status !== AnalysisStatus.failed).toBeTruthy();
      } while (response.value.status !== AnalysisStatus.complete);

      expect(response.value.sarif.runs[0].results).toHaveLength(4);
    },
    TEST_TIMEOUT,
  );
});
