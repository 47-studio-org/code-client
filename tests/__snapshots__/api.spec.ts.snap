// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Requests to public API test successful workflow with and without linters 1`] = `
Object {
  "2": Array [
    Object {
      "cols": Array [
        8,
        27,
      ],
      "fingerprints": Array [
        Object {
          "fingerprint": "f8e3391465a47f6586489cffd1f44ae47a1c4885c722de596d6eb931fe43bb16",
          "version": 0,
        },
      ],
      "markers": Array [],
      "rows": Array [
        5,
        5,
      ],
    },
  ],
  "3": Array [
    Object {
      "cols": Array [
        6,
        25,
      ],
      "fingerprints": Array [
        Object {
          "fingerprint": "3e40a81739245db8fff4903a7e28e08bffa03486a677e7c91594cfdf15fb5a1d",
          "version": 0,
        },
        Object {
          "fingerprint": "57664a44.2c254dac.98501263.9e345555.da547a36.9509b717.a713c1c8.45d76bdf.57664a44.2c254dac.98501263.9e345555.da547a36.9509b717.a713c1c8.45d76bdf",
          "version": 1,
        },
      ],
      "markers": Array [
        Object {
          "msg": Array [
            25,
            36,
          ],
          "pos": Array [
            Object {
              "cols": Array [
                7,
                14,
              ],
              "file": "/AnnotatorTest.cpp",
              "rows": Array [
                8,
                8,
              ],
            },
          ],
        },
        Object {
          "msg": Array [
            45,
            57,
          ],
          "pos": Array [
            Object {
              "cols": Array [
                6,
                25,
              ],
              "file": "/AnnotatorTest.cpp",
              "rows": Array [
                10,
                10,
              ],
            },
          ],
        },
      ],
      "rows": Array [
        10,
        10,
      ],
    },
  ],
}
`;
